From 54721b2dea5f5dc66f131838196a47e08b028791 Mon Sep 17 00:00:00 2001
From: Kuninori Morimoto <kuninori.morimoto.gx@renesas.com>
Date: Wed, 12 Jun 2013 11:32:01 +0900
Subject: [PATCH 495/675] LTSI: ARM: shmobile: r8a7778: setup early timer

LTSI v3.4 ARM kernel doesn't support .init_time on DT_MACHINE_START,
but, upstream base r8a7778 SoC support is using it.
The problem is that the early timer initialization method between
v3.4 and latest kernel is different.

Signed-off-by: Kuninori Morimoto <kuninori.morimoto.gx@renesas.com>
---
 arch/arm/mach-shmobile/setup-r8a7778.c | 41 ++++++++++++++++++++++++++++++----
 1 file changed, 37 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index 57d6b0e..d5d1b67 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -31,6 +31,7 @@
 #include <mach/r8a7778.h>
 #include <mach/common.h>
 #include <asm/mach/arch.h>
+#include <asm/mach/time.h>
 #include <asm/hardware/cache-l2x0.h>
 
 /* SCIF */
@@ -105,9 +106,6 @@ void __init r8a7778_add_standard_devices(void)
 		platform_device_register_data(&platform_bus, "sh-sci", i,
 					      &scif_platform_data[i],
 					      sizeof(struct plat_sci_port));
-
-	r8a7778_register_tmu(0);
-	r8a7778_register_tmu(1);
 }
 
 #define INT2SMSKCR0	0x82288 /* 0xfe782288 */
@@ -147,9 +145,44 @@ void __init r8a7778_init_irq(void)
 	r8a7778_init_irq_common();
 }
 
+static void __init r8a7778_earlytimer_init(void)
+{
+	r8a7778_clock_init();
+	shmobile_earlytimer_init();
+}
+
+static struct platform_device sh_tmu0_device = {
+	.name		= "sh_tmu",
+	.id		= 0,
+	.dev = {
+		.platform_data	= &sh_tmu0_platform_data,
+	},
+	.resource	= sh_tmu0_resources,
+	.num_resources	= ARRAY_SIZE(sh_tmu0_resources),
+};
+
+static struct platform_device sh_tmu1_device = {
+	.name		= "sh_tmu",
+	.id		= 1,
+	.dev = {
+		.platform_data	= &sh_tmu1_platform_data,
+	},
+	.resource	= sh_tmu1_resources,
+	.num_resources	= ARRAY_SIZE(sh_tmu1_resources),
+};
+
+static struct platform_device *r8a7778_early_devices[] __initdata = {
+	&sh_tmu0_device,
+	&sh_tmu1_device,
+};
+
 void __init r8a7778_init_delay(void)
 {
+	early_platform_add_devices(r8a7778_early_devices,
+				   ARRAY_SIZE(r8a7778_early_devices));
+
 	shmobile_setup_delay(800, 1, 3); /* Cortex-A9 @ 800MHz */
+	shmobile_timer.init = r8a7778_earlytimer_init;
 }
 
 #ifdef CONFIG_USE_OF
@@ -178,7 +211,7 @@ DT_MACHINE_START(R8A7778_DT, "Generic R8A7778 (Flattened Device Tree)")
 	.init_early	= r8a7778_init_delay,
 	.init_irq	= r8a7778_init_irq_dt,
 	.init_machine	= r8a7778_add_standard_devices_dt,
-	.init_time	= shmobile_timer_init,
+	.timer		= &shmobile_timer,
 	.dt_compat	= r8a7778_compat_dt,
 MACHINE_END
 
-- 
1.8.2

